processout-sdk/src/main/java/com/processout/processout_sdk/ProcessOut.java

```diff
@@ -82,59 +97,204 @@ public void updateCvc(Card card, final CvcUpdateCallback callback) {
+                        } catch (JSONException e) {
+                            e.printStackTrace();
+                            callback.onError(e);
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Allow card payments authorization (with 3DS2 support)
+     *
+     * @param invoiceId previously generated invoice
+     * @param source    source to use for the charge (card token, etc.)
+     * @param handler   (Custom 3DS2 handler)
+     */
+    public void makeCardPayment(final String invoiceId, final String source, final ThreeDSHandler handler) {
+        try {
+            // Generate the authorization body and forces 3DS2
+            AuthorizationRequest authRequest = new AuthorizationRequest(source);
+            final JSONObject body = new JSONObject(new Gson().toJson(authRequest));
+
+            Network.getInstance(this.context, this.projectId).CallProcessOut(
+                    "/invoices/" + invoiceId + "/authorize", Request.Method.POST,
+                    body, new Network.NetworkResult() {
+                        @Override
+                        public void onError(Exception error) {
+                            handler.onError(error);
+                        }
+
+                        @Override
+                        public void onSuccess(JSONObject json) {
+                            Gson gson = new Gson();
+                            // Handle the authorization result
+                            AuthorizationResult result = gson.fromJson(
+                                    json.toString(), AuthorizationResult.class);
+                            handleAuthorizationResult(invoiceId, source, handler, result);
+                        }
+                    });
+        } catch (JSONException e) {
+            handler.onError(e);
+        }
+    }
+
+    private void handleAuthorizationResult(
+            final String invoiceId, final String source, final ThreeDSHandler handler, AuthorizationResult result) {
+        CustomerAction cA = result.getCustomerAction();
+        if (cA == null) {
+            // No customer action in the authorization result, we return the invoice id
+            handler.onSuccess(invoiceId);
+            return;
+        }
+
+        // Customer action required
+        Type mapType = new TypeToken<Map<String, String>>() {
+        }.getType();
+        switch (cA.getType()) {
+            case FINGERPRINT_MOBILE:
+                DirectoryServerData directoryServerData = new Gson().fromJson(new String(Base64.decode(cA.getValue().getBytes(), Base64.NO_WRAP)), DirectoryServerData.class);
+                handler.doFingerprint(directoryServerData, new ThreeDSHandler.DoFingerprintCallback() {
+                    @Override
+                    public void continueCallback(ThreeDSFingerprintResponse request) {
+                        Gson gson = new GsonBuilder().disableHtmlEscaping().create();
```

Comment:
Should we maybe store this within our object instead of re-init it at every execution?